app/master/build.py

```diff
@@ -327,13 +332,17 @@ def cancel(self):
         self._record_state_timestamp(BuildStatus.CANCELED)
 
         # Deplete the unstarted subjob queue.
-        # TODO: Handle situation where cancel() is called while subjobs are being added to _unstarted_subjobs
-        while self._unstarted_subjobs is not None and not self._unstarted_subjobs.empty():
-            try:
-                # A subjob may be asynchronously pulled from this queue, so we need to avoid blocking when empty.
-                self._unstarted_subjobs.get(block=False)
-            except Empty:
-                break
+        with self._subjob_queuing_lock:
+            while self._unstarted_subjobs is not None and not self._unstarted_subjobs.empty():
+                try:
+                    # A subjob may be asynchronously pulled from this queue, so we need to avoid blocking when empty.
+                    self._unstarted_subjobs.get(block=False)
+                except Empty:
+                    break
+
+        # Tell each allocated slave to teardown
+        for slave in self._slaves_allocated:
+            self.execute_next_subjob_or_teardown_slave(slave)
```

Comment:
I'm not sure it will be as easy as calling `execute_next_subjob_or_teardown_slave` for each slave. It's kind of misleading -- a better name for that method would probably be `execute_next_subjob_or_free_executor`.

When cancel is called, there may currently be subjobs executing on the slave (in our case, up to 10). I think in that case, calling `execute_next_subjob_or_teardown_slave` for each slave will just make the master think there are now 9 busy executors on each slave. In fact if all 10 of those subjobs then report results back, the last one will probably cause a request exception since we'll try to free 1 executor when the master thinks all 10 were already freed.